File: spec/models/query_string_parser_spec.rb
require "spec_helper"

describe QueryStringParser do
  it "can get a single pair of arguments" do
    expect(QueryStringParser.new('http://www.example.com?baby=boomba').query_hash).to eq({'baby' => 'boomba'})
  end

  it "can get a multiple arguments" do
    expect(QueryStringParser.new('http://www.example.com?baby=boomba&cookie=sugar').query_hash).to eq({'baby' => 'boomba', 'cookie' => 'sugar'})
  end

  it "handle no query string" do
    expect(QueryStringParser.new('http://www.example.com').query_hash).to eq({})
  end

  it "can handle two = signs without a seperating &" do
    expect(QueryStringParser.new('http://www.example.com?poop=pam=pamper').query_hash).to eq({'poop' => 'pam=pamper'})
  end

  it "can handle no argument queries" do
    expect(QueryStringParser.new('http://www.example.com?poop&pam=power&ping=').query_hash).to eq({'poop' => nil, 'pam' => "power", 'ping' => ""})
  end

  it "can turn a + into a space" do
    expect(QueryStringParser.new('http://www.example.com?value=boom+bam').query_hash).to eq({'value' => 'boom bam'})
  end

  it "can decode %26" do
    expect(QueryStringParser.new('http://www.example.com?value=boom+%26+bam').query_hash).to eq({'value' => 'boom & bam'})
  end

  it "can decode keys as well" do
    expect(QueryStringParser.new('http://www.example.com?boom+%26+bam=value').query_hash).to eq({'boom & bam' => 'value'})
  end

  it "can turn the same keys into an array" do
    expect(QueryStringParser.new('http://www.example.com?key=value1&key=value2&key=value3').query_hash).to eq({'key' => ['value1', 'value2', 'value3']})
  end

  it "can ignore an empty value" do
    expect(QueryStringParser.new('http://www.example.com?a&&b').query_hash).to eq({'a' => nil, 'b' => nil})
  end

  it "can ignore empty keys" do
    expect(QueryStringParser.new('http://www.example.com?pow&=pow2').query_hash).to eq({'pow' => nil})
  end

end

File: code/js/spec/query_string_parser_spec.js
describe("QueryStringParser", function() {
    
    it("can get a single pair of arguments", function() {
        expect((new QueryStringParser('http://www.example.com?baby=boomba')).queryHash()).toEqual({'baby': 'boomba'})
    });

    it("can get a multiple arguments", function() {
        expect((new QueryStringParser('http://www.example.com?baby=boomba&cookie=sugar')).queryHash()).toEqual({'baby':  'boomba', 'cookie':  'sugar'})
    });

    it("handle no query string", function() {
        expect((new QueryStringParser('http://www.example.com')).queryHash()).toEqual({})
    });

    it("can handle two = signs without a seperating &", function() {
        expect((new QueryStringParser('http://www.example.com?poop=pam=pamper')).queryHash()).toEqual({'poop':  'pam=pamper'})
    });

    it("can handle no argument queries", function() {
        expect((new QueryStringParser('http://www.example.com?poop&pam=power&ping=')).queryHash()).toEqual({'poop':  null, 'pam':  "power", 'ping':  ""})
    });

    it("can turn a + into a space", function() {
        expect((new QueryStringParser('http://www.example.com?value=boom+bam')).queryHash()).toEqual({'value':  'boom bam'})
    });

    it("can decode %26", function() {
        expect((new QueryStringParser('http://www.example.com?value=boom+%26+bam')).queryHash()).toEqual({'value':  'boom & bam'})
    });

    it("can decode keys as well", function() {
        expect((new QueryStringParser('http://www.example.com?boom+%26+bam=value')).queryHash()).toEqual({'boom & bam':  'value'})
    });

    it("can turn the same keys into an array", function() {
        expect((new QueryStringParser('http://www.example.com?key=value1&key=value2&key=value3')).queryHash()).toEqual({'key':  ['value1', 'value2', 'value3']})
    });

    it("can ignore an empty value", function() {
        expect((new QueryStringParser('http://www.example.com?a&&b')).queryHash()).toEqual({'a':  null, 'b':  null})
    });

    it("can ignore empty keys", function() {
        expect((new QueryStringParser('http://www.example.com?pow&=pow2')).queryHash()).toEqual({'pow':  null})
    });
});
